### PR TITLE
Refactor Inventory client

### PR DESCRIPTION
**Pull request by rquitales**
This is a WIP PR pending rebase when K8s resources are updated to use v1.25.

This PR uses Karl's initial PR as the basis for the Inventory client refactor.

- Rebased Karl's code onto cli-utils tip
- Updated deepcopy-gen to also create `DeepCopyObject` for actuation.Inventory
- Remove stuttering on certain Inventory objects and methods after rebase

TODO: Rebase PR when blocking PRs are merged

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -19,7 +19,7 @@ all: generate license fix vet fmt test lint tidy
 	go install github.com/golangci/golangci-lint/cmd/golangci-lint@v1.44.2
 
 "$(MYGOBIN)/deepcopy-gen":
-	go install k8s.io/code-generator/cmd/deepcopy-gen@v0.23.6
+	go install k8s.io/code-generator/cmd/deepcopy-gen@v0.24.1
```
Comment (karlkfi) [inline Makefile:22]: is this needed for this PR or just coincidental? It can probably be pulled out.

File: Makefile [hunk 2]
```diff
@@ -63,7 +63,7 @@ install-column-apt:
 
 .PHONY: generate-deepcopy
 generate-deepcopy: "$(MYGOBIN)/deepcopy-gen"
-	hack/run-in-gopath.sh deepcopy-gen --input-dirs ./pkg/apis/... -O zz_generated.deepcopy --go-header-file ./LICENSE_TEMPLATE_GO
+	hack/run-in-gopath.sh deepcopy-gen --input-dirs ./pkg/apis/actuation/ -O zz_generated.deepcopy --go-header-file ./LICENSE_TEMPLATE_GO
```
Comment (karlkfi) [inline Makefile:66]: is this needed for this PR? It can probably be pulled out.

File: pkg/apis/actuation/types.go [hunk 3]
```diff
@@ -8,6 +8,8 @@ import (
 	"k8s.io/apimachinery/pkg/types"
 )
 
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
```
Comment (karlkfi) [inline pkg/apis/actuation/types.go:11]: is this needed for this PR? It can probably be pulled out.

File: pkg/inventory/client-factory.go [hunk 4]
```diff
@@ -0,0 +1,11 @@
+// Copyright 2021 The Kubernetes Authors.
```
Comment (janetkuo) [inline pkg/inventory/client-factory.go:1]: nit: 2022 (and other files too)

File: pkg/apply/common_test.go [hunk 5]
```diff
@@ -27,55 +27,83 @@ import (
...
+		Name:      inventoryObj.GetName(),
+		Namespace: inventoryObj.GetNamespace(),
+		Labels:    inventoryObj.GetLabels(),
+	}
+)
 
-func (i inventoryInfo) toWrapped() inventory.Info {
-	return inventory.WrapInventoryInfoObj(i.toUnstructured())
-}
+// type inventoryInfo struct {
```
Comment (Liujingfang1) [inline pkg/apply/common_test.go:66]: Looks like this type is not needed any more. You can remove it.

File: cmd/main.go [hunk 6]
```diff
@@ -62,7 +62,8 @@ func main() {
 	}
 
 	loader := manifestreader.NewManifestLoader(f)
-	invFactory := inventory.ClusterClientFactory{StatusPolicy: inventory.StatusPolicyNone}
+	// TODO: Add StatusPolicy back to factory.
```
Comment (karlkfi) [inline cmd/main.go:65]: Does ConfigMapClientFactory need configurable StatusPolicy? Or is it configured elsewhere?

File: cmd/status/cmdstatus.go [hunk 7]
```diff
@@ -164,11 +164,14 @@ func (r *Runner) loadInvFromDisk(cmd *cobra.Command, args []string) (*printer.Pr
 
 	// Based on the inventory template manifest we look up the inventory
 	// from the live state using the inventory client.
-	identifiers, err := invClient.GetClusterObjs(inv)
+	inv, err := invClient.Load(context.TODO(), invInfo)
```
Comment (karlkfi) [inline cmd/status/cmdstatus.go:167]: Does the cobra.Command have a context we can pass in here?

Comment (karlkfi) [reply cmd/status/cmdstatus.go:167]: `r.ctx` is used elsewhere

Comment (rquitales) [reply cmd/status/cmdstatus.go]: Ah yes. We can re-use that. Done.

File: cmd/main.go (hunk 6 above)
Comment (rquitales) [reply cmd/main.go]: ConfigMapClientFactory does need configurable StatusPolicy. We have historically hardcoded `StatusPolicyNone` event when we have a configurable field for `kapply`. Perhaps we should also add a flag to `kapply` to have users set this?

Regardless, ConfigMap now has a configurable StatusPolicy field for handling status conversion to the data fields.

File: pkg/apis/actuation/types.go (hunk 3 above)
Comment (rquitales) [reply pkg/apis/actuation/types.go:11]: The refactor now requires `Inventory` to satisfy `runtime.Object`  which was previously not needed, hence why deepcopy-gen tags are added in this PR. We could spin it out into another PR that is merged first, but this tag addition is related to the refactor. LMKYT

File: Makefile (hunk 2 above)
Comment (rquitales) [reply Makefile:66]: No longer needed.

File: Makefile (hunk 1 above)
Comment (rquitales) [reply Makefile:22]: deepcopy-gen needed to be bumped to generate the right methods for `Inventory` to satisfy `runtime.Object` interface. This will be handled in #606 and this PR will rebase to pull this change out.

File: pkg/apply/task/inv_set_task_test.go [hunk 8]
```diff
@@ -164,18 +166,25 @@ func TestInvSetTask(t *testing.T) {
 
 	for name, tc := range tests {
 		t.Run(name, func(t *testing.T) {
-			client := inventory.NewFakeClient(object.ObjMetadataSet{})
+			// client := inventory.NewFakeInventoryClient(object.ObjMetadataSet{})
```
Comment (karlkfi) [inline pkg/apply/task/inv_set_task_test.go:169]: remove commented line

File: test/e2e/e2e_test.go [hunk 9]
```diff
@@ -29,7 +29,7 @@ const (
 
 var inventoryConfigs = map[string]invconfig.InventoryConfig{}
 var inventoryConfigTypes = []string{
-	ConfigMapTypeInvConfig,
+	// ConfigMapTypeInvConfig,
```
Comment (karlkfi) [inline test/e2e/e2e_test.go:32]: why commented out?

File: pkg/apply/destroyer.go [hunk 10]
```diff
@@ -71,6 +72,43 @@ func setDestroyerDefaults(o *DestroyerOptions) {
 	}
 }
 
+// prepareObjects returns the set of objects to prune or an error if one occurred.
+func (d *Destroyer) prepareObjects(
```
Comment (karlkfi) [inline pkg/apply/destroyer.go:76]: I wonder if this refactor, to move the Pruner.GetPruneObjs code into Destoryer.prepareObjects could be done by itself in a separate PR. It would have to go first, and it wouldn't look exactly like this, because it would use the old inventory info and client, but it would at least move the code into the right place, which would make the diff easier to follow in this PR.

Does that sound worth the extra effort?

File: pkg/apply/destroyer_test.go [hunk 11]
```diff
@@ -43,13 +45,10 @@ func TestDestroyerCancel(t *testing.T) {
...
 			runTimeout:       2 * time.Second,
 			testTimeout:      30 * time.Second,
-			invInfo: inventoryInfo{
-				name:      "abc-123",
-				namespace: "test",
-				id:        "test",
-				set: object.ObjMetadataSet{
-					testutil.ToIdentifier(t, resources["deployment"]),
-				},
+			invObjs: object.ObjMetadataSet{
```
Comment (karlkfi) [inline pkg/apply/destroyer_test.go:48]: This minor refactor could be done in a separate PR. It would just modify this test to build the inventoryInfo in the test body, instead of here in the testCases.

File: pkg/apply/prune/prune.go [hunk 12]
```diff
@@ -200,58 +198,18 @@ func (p *Pruner) removeInventoryAnnotation(obj *unstructured.Unstructured) (*uns
...
-		if _, ok := annotations[inventory.OwningInventoryKey]; ok {
-			klog.V(4).Infof("removing annotation (object: %q, annotation: %q)", id, inventory.OwningInventoryKey)
-			delete(annotations, inventory.OwningInventoryKey)
-			obj.SetAnnotations(annotations)
-			namespacedClient, err := p.namespacedClient(id)
-			if err != nil {
-				return obj, err
-			}
-			_, err = namespacedClient.Update(context.TODO(), obj, metav1.UpdateOptions{})
+	if inventory.OwningInventoryAnnotation(obj) != "" {
```
Comment (karlkfi) [inline pkg/apply/prune/prune.go:201]: The addition and use of the OwningInventoryAnnotation and DeleteOwningInventoryAnnotation functions could be moved to its own PR. It's just a cleanup, not dependent on the inv client changes.

File: pkg/apply/prune/prune_test.go [hunk 13]
```diff
@@ -614,21 +589,21 @@ func TestPrune(t *testing.T) {
 			for _, id := range tc.expectedFailed {
 				assert.Truef(t, im.IsFailedDelete(id), "Prune() should mark object as failed: %s", id)
 			}
-			for _, id := range pruneIds.Diff(tc.expectedFailed) {
+			for _, id := range pruneIDs.Diff(tc.expectedFailed) {
```
Comment (karlkfi) [inline pkg/apply/prune/prune_test.go:592]: pruneIds -> pruneIDs could be pulled out to a smaller refactor/rename PR.

File: pkg/apply/prune/prune_test.go [hunk 14]
```diff
@@ -784,88 +776,12 @@ func TestPruneWithErrors(t *testing.T) {
...
-				t.Fatalf("expected %d prune objs, got %d", len(tc.expectedObjs), len(actualObjs))
-			}
-			actualIds := object.UnstructuredSetToObjMetadataSet(actualObjs)
-			expectedIds := object.UnstructuredSetToObjMetadataSet(tc.expectedObjs)
-			if !object.ObjMetadataSetEquals(expectedIds, actualIds) {
-				t.Errorf("expected prune objects (%v), got (%v)", expectedIds, actualIds)
+				events = append(events, e)
 			}
+			receivedEvents := testutil.EventsToExpEvents(events)
+			testutil.AssertEqual(t, tc.expectedEvents, receivedEvents)
```
Comment (karlkfi) [inline pkg/apply/prune/prune_test.go:784]: This test cleanup to use testutil.AssertEqual and testutil.EqualErrorString could be pulled out into another PR.

File: pkg/apply/prune/prune_test.go [hunk 15]
```diff
@@ -973,7 +888,7 @@ func TestPrune_PropagationPolicy(t *testing.T) {
 			eventChannel := make(chan event.Event, 1)
 			resourceCache := cache.NewResourceCacheMap()
 			taskContext := taskrunner.NewTaskContext(eventChannel, resourceCache)
-			err := po.Prune([]*unstructured.Unstructured{pdb}, []filter.ValidationFilter{}, taskContext, "test-0", Options{
+			err := po.Prune(object.UnstructuredSet{pdb}, []filter.ValidationFilter{}, taskContext, "test-0", Options{
```
Comment (karlkfi) [inline pkg/apply/prune/prune_test.go:891]: This minor cleanup could be in a separate PR, maybe with other cleanups in this test file.

File: pkg/apply/task/delete_inv_task.go [hunk 16]
```diff
@@ -40,13 +41,66 @@ func (i *DeleteInvTask) Identifiers() object.ObjMetadataSet {
...
+				err = fmt.Errorf("failed to update inventory: %w", err)
+				i.sendTaskResult(taskContext, err)
+				return
+			}
+
+			i.sendTaskResult(taskContext, nil)
+			return
+		}
+
+		invInfo := inventory.InfoFromObject(im.Inventory())
```
Comment (karlkfi) [inline pkg/apply/task/delete_inv_task.go:95]: I think all the changes in this file above this line are probably an enhancement to make the inventory not be deleted when there are delete errors. I think all that can be pulled out to a follow up PR. I don't think it's required for the client rewrite. But you'll have to make sure the tests pass without it to really know. Some of the test changes probably also need to go with this.

File: pkg/apply/task/delete_inv_task_test.go [hunk 17]
```diff
@@ -4,65 +4,144 @@
...
 	"sigs.k8s.io/cli-utils/pkg/apply/taskrunner"
 	"sigs.k8s.io/cli-utils/pkg/common"
 	"sigs.k8s.io/cli-utils/pkg/inventory"
-	"sigs.k8s.io/cli-utils/pkg/object"
+	"sigs.k8s.io/cli-utils/pkg/testutil"
 )
 
 func TestDeleteInvTask(t *testing.T) {
+	ref1 := inventory.ObjectReferenceFromObject(obj1)
+	// ref2 := inventory.ObjectReferenceFromObject(obj2)
```
Comment (karlkfi) [inline pkg/apply/task/delete_inv_task_test.go:23]: i think these comments lines can be removed.

File: pkg/apply/task/delete_inv_task_test.go [hunk 18]
```diff
@@ -4,65 +4,144 @@
...
-					t.Errorf("running DeleteInvTask expected error (%s), got (%s)", tc.err, result.Err)
-				}
-			} else {
-				if result.Err != nil {
-					t.Errorf("unexpected error running DeleteInvTask: %s", result.Err)
-				}
+			task.Start(taskContext)
+			result := <-taskContext.TaskChannel()
+			if tc.expectedError != nil {
+				assert.EqualError(t, result.Err, tc.expectedError.Error())
```
Comment (karlkfi) [inline pkg/apply/task/delete_inv_task_test.go:137]: It might be possible to pull out this assert.EqualError cleanup to its own PR. It can hopefully be done separately from the subsequent testutil.AssertEqual of the inventory after Load. But given the high amount of other test changes here, it's not really a big deal to pull out unless there are other similar changes in other test files.

File: pkg/apply/task/inv_add_task.go [hunk 19]
```diff
@@ -38,30 +49,82 @@ func (i *InvAddTask) Action() event.ResourceAction {
 }
 
 func (i *InvAddTask) Identifiers() object.ObjMetadataSet {
-	return object.UnstructuredSetToObjMetadataSet(i.Objects)
+	return object.UnstructuredSetToObjMetadataSet(i.ApplyObjects).Union(
```
Comment (karlkfi) [inline pkg/apply/task/inv_add_task.go:52]: This refactor to pass in ApplyObjects and PruneObjects, instead of just Objects might be extractable. But I'm not sure if it makes sense by itself or not.

File: pkg/apply/task/inv_add_task.go [hunk 20]
```diff
@@ -71,25 +134,51 @@ func (i *InvAddTask) Cancel(_ *taskrunner.TaskContext) {}
...
+// NamespaceInSet returns the the namespace the passed inventory
 // object will be applied to, or nil if this namespace object does not exist
 // in the passed slice "infos" or the inventory object is cluster-scoped.
-func inventoryNamespaceInSet(inv inventory.Info, objs object.UnstructuredSet) *unstructured.Unstructured {
-	if inv == nil {
-		return nil
-	}
-	invNamespace := inv.Namespace()
-
+func NamespaceInSet(inv client.Object, objs object.UnstructuredSet) *unstructured.Unstructured {
```
Comment (karlkfi) [inline pkg/apply/task/inv_add_task.go:140]: inventoryNamespaceInSet -> NamespaceInSet rename could be pulled out to a rename/refactor PR.

File: pkg/apply/task/inv_add_task_test.go [hunk 21]
```diff
@@ -72,110 +78,164 @@ func TestInvAddTask(t *testing.T) {
...
-			actual, _ := client.GetClusterObjs(nil)
-			if !tc.expectedObjs.Equal(actual) {
-				t.Errorf("expected merged inventory (%s), got (%s)", tc.expectedObjs, actual)
+			assert.Equal(t, taskName, task.Name())
+			testutil.AssertEqual(t, tc.expectedIds, task.Identifiers())
+
+			task.Start(taskContext)
+			result := <-taskContext.TaskChannel()
+			if tc.expectedError != nil {
+				assert.EqualError(t, result.Err, tc.expectedError.Error())
```
Comment (karlkfi) [inline pkg/apply/task/inv_add_task_test.go:188]: Here's another assert.EqualError cleanup that could be bundled with others in another PR.

File: pkg/apply/task/inv_add_task_test.go [hunk 22]
```diff
@@ -72,110 +78,164 @@ func TestInvAddTask(t *testing.T) {
...
+				return
 			}
+			assert.NoError(t, result.Err)
+
+			inv, err := client.Load(context.TODO(), inventoryInfo)
+			assert.NoError(t, err)
+			testutil.AssertEqual(t, tc.expectedSpecObjs, inv.Spec.Objects)
```
Comment (karlkfi) [inline pkg/apply/task/inv_add_task_test.go:195]: `!tc.expectedObjs.Equal(actual)` -> `testutil.AssertEqual(t, tc.expectedSpecObjs, inv.Spec.Objects)` seems like an independent test cleanup.

The subsequent test for expectedStatusObjs is new, but the spec/actual validation could be done with the other GetClusterObjs in a separate test cleanup PR.

File: pkg/apply/task/inv_set_task_test.go [hunk 23]
```diff
@@ -164,18 +166,25 @@ func TestInvSetTask(t *testing.T) {
...
+			client := &inventory.InMemoryClient{}
```
Comment (karlkfi) [inline pkg/apply/task/inv_set_task_test.go:170]: Does it make sense to replace the `inventory.NewFakeClient` with the `inventory.InMemoryClient` before changing the client interface? Or would that just be wasted effort?

File: pkg/inventory/configmap-converter.go [hunk 24]
```diff
@@ -0,0 +1,126 @@
...
+	}
+	objs, err := object.FromStringMap(data)
+	if err != nil {
+		return nil, err
+	}
+	inv.Spec.Objects = ObjectReferencesFromObjMetadataSet(objs)
+	// Sort objects to reduce chun on update.
+	sort.Sort(AlphanumericObjectReferences(inv.Spec.Objects))
+
+	// // Convert in.Data to out.status.objects
```
Comment (karlkfi) [inline pkg/inventory/configmap-converter.go:58]: Are we missing some code here?

File: pkg/inventory/configmap-converter.go [hunk 25]
```diff
@@ -0,0 +1,126 @@
...
+			objMap[objMetadata.String()] = stringFrom(status)
+		} else {
+			// It's possible that the passed in status doesn't any object status
+			objMap[objMetadata.String()] = ""
+		}
+	}
+	return objMap
+}
+
+func stringFrom(status actuation.ObjectStatus) string {
```
Comment (karlkfi) [inline pkg/inventory/configmap-converter.go:115]: what's this func do? is its serializing the status? please add a comment and maybe rename for clarity.

Is there also a parallel deserialize function? If so, maybe marshal and unmarshal? or serialize and deserialize?

File: pkg/inventory/fake-client-factory.go [hunk 26]
```diff
@@ -0,0 +1,17 @@
...
+// SPDX-License-Identifier: Apache-2.0
+
+package inventory
+
+import (
+	cmdutil "k8s.io/kubectl/pkg/cmd/util"
+	"sigs.k8s.io/cli-utils/pkg/object"
+)
+
+type FakeClientFactory object.ObjMetadataSet
```
Comment (karlkfi) [inline pkg/inventory/fake-client-factory.go:11]: public types need comments. Not sure why the linter didn't catch this.

File: pkg/inventory/inventory_error.go [hunk 27]
```diff
@@ -25,41 +25,31 @@ const multipleInventoryErrorStr = `Package has multiple inventory object templat
...
 
+const namespaceInSetErrorStr = `Inventory use namespace defined in package.
+
+The inventory cannot use a namespace that is defined in the package.
+`
+
 type NoInventoryObjError struct{}
 
-func (e *NoInventoryObjError) Error() string {
+func (g NoInventoryObjError) Error() string {
```
Comment (karlkfi) [inline pkg/inventory/inventory_error.go:35]: Generally, error types should use a pointer receiver, to allow them to be nil, even if there's no state that actually needs a pointer. Making the Error method require a pointer is a good way to force users to follow this best practice when using it.

File: pkg/inventory/inventory_test.go [hunk 28]
```diff
@@ -234,18 +204,8 @@ func TestRetrieveInventoryLabel(t *testing.T) {
...
-		if !test.isError {
-			if err != nil {
-				t.Fatalf("Received unexpected error: %s\n", err)
-			}
-			if test.inventoryLabel != actual {
-				t.Errorf("Expected inventory label (%s), got (%s)\n", test.inventoryLabel, actual)
-			}
-		}
+		actual := Label(object.InfoToUnstructured(test.inventoryInfo))
+		assert.Equal(t, test.inventoryLabel, actual)
```
Comment (karlkfi) [inline pkg/inventory/inventory_test.go:208]: Here's another assert.Equal test cleanup that could be pulled out.

File: pkg/inventory/main_test.go [hunk 29]
```diff
@@ -0,0 +1,19 @@
+// Copyright 2022 The Kubernetes Authors.
```
Comment (karlkfi) [inline pkg/inventory/main_test.go:1]: This test file could be pulled out. It's not a big deal if there's only one, but it's completely independent. It just allows the package to be run with logging verbosity.

File: pkg/inventory/type-conv.go [hunk 30]
```diff
@@ -4,11 +4,22 @@
...
+	"fmt"
+
 	"k8s.io/apimachinery/pkg/runtime/schema"
 	"sigs.k8s.io/cli-utils/pkg/apis/actuation"
 	"sigs.k8s.io/cli-utils/pkg/object"
+	"sigs.k8s.io/controller-runtime/pkg/client"
 )
 
+// ObjMetadataEqualObjectReference compares an ObjMetadata with a ObjectReference
+func ObjMetadataEqualObjectReference(id object.ObjMetadata, ref actuation.ObjectReference) bool {
```
Comment (karlkfi) [inline pkg/inventory/type-conv.go:16]: This helper function looks pretty independent. It could be in a separate PR.

In fact, this whole file seems like helper funcs that could be factored into a separate PR.

**Review by karlkfi — CHANGES_REQUESTED** [review]
I've identified a bunch of code that looks independent to the core of this PR (to rewrite the inventory client). They don't ALL need to be pulled out, and they don't ALL need to be individual PRs. Feel free to group them at your own discretion. Then link them from this PR and assign them to me. After they're merged, you can rebase this PR.
